src/python/WMCore/MicroService/MSPileup/MSPileupTasks.py

```diff
@@ -100,6 +104,95 @@ def pileupSizeTask(self):
+
+            # check previous fraction value and proceed only if has changed
+            lastFraction = 0
+            transition = doc.get('transition', [])
+            if transition:
+                lastFraction = transition[-1]['containerFraction']
+            if lastFraction == fraction:
+                # do nothing since there is no transition, i.e. skip this record
+                continue
+
+            # get rucio DIDs (datasets, in CMS name blocks) for our pileup document as following:
+            # - when container fraction is decreasing: new container is a subset of customName
+            # - when container fraction is increasing: new container is a superset of customName + subset of pileupName
+
+            # by default we assume it is decreasing
+            pname = doc['customName']
+            blockNames = self.rucioClient.getBlocksInContainer(pname)
+
+            # but if it increasing
+            if lastFraction < fraction:
+                pname = doc['pileupName']
+                blockNames += self.rucioClient.getBlocksInContainer(pname)
+
+            # get portion of DIDs based on ceil(containerFraction * num_rucio_datasets)
+            portion = blockNames[:math.ceil(fraction * len(blockNames))]
+
+            self.logger.info("For pileup document '%s' take portion '%s'", doc, blockNames)
+
+            # create new container DID in Rucio
+            for blockName in portion:
+                self.rucioClient.createContainer(blockName, scope=self.customRucioScope)
+
+            # add new transition record
+            transition = doc['transition']
+            tdoc = transition[-1]
+            trec = {'customDID': tdoc['customDID'], 'updateTime': gmtimeSeconds(),
+                    'containerFraction': tdoc['containerFraction'], 'DN': tdoc['DN']}
+            transition.append(trec)
+            doc['transition'] = transition
+
+            # we call attachDIDs Rucio wrapper API with our set of DIDs
+            for rse in doc['currentRSEs']:
+                # superDID upper structure level (can be a container or block. If attaching blocks,
+                # then it's a container name
+                superDID = portion[0].split('#')[0]
+                self.rucioClient.attachDIDs(rse, superDID, portion, scope=self.customRucioScope)
+
+            # create new rules for custom DID using provide by rseExpression
+            newRules = self.rucioClient.createReplicationRules(portion, rseExpression)
+
+            # set expiration date (to be 24h) for already existing ruleIds from pileup document
+            for rid in doc['ruleIds']:
+                # set expiration date to be 24h ahead of right now
+                opts = {'lifetime': 24 * 60 * 60}
+                self.rucioClient.updateRule(rid, opts)
+
+            # add new ruleIds to pileup document
+            doc['ruleIds'] += newRules
+
+            self.logger.info("partialPileupTask updated %s", doc)
```

Comment:
I would suggest to rephrase this to something like "Pileup name had its fraction updated in the partialPileupTask function".

Reply:
ok